File: reactionBalancer_ITA.py
#            88         
#            88         
#            88         
# ,adPPYba,  88   ,d8   
#a8"     "8a 88 ,a8"    
#8b       d8 8888[      
#"8a,   ,a8" 88`"Yba,   
# `"YbbdP"'  88   `Y8a  

#--Variabili e librerie utili--#
import random
Stato = 'n' #Variabile che userò per determinare la fine di un ciclo
Contatore1 = 0 #Variabile che userò per contare i cicli
Contatore2 = 0 #Un'altra variabile che userò per contare cicli che avvengo in contemporanea a quelli contati da "Contatore1"

##-------------------------Prelievo input-----------------------------##
#--Determinazione degli elementi della reazione--#
Elementi = [] #Lista degli elementi che reagiranno
while Stato != 'S': #Ciclo di registrazione degli elementi coinvolti
	Elemento = input("Quale elemento è presente in questa equazione?")
	Elementi += [Elemento]
	Stato = input("Questo era l'ultimo elemento? (S/n)")
print("Ricapitolando, gli elementi che hai elencato sono:")
print(Elementi)

#--Assegnazione degli atomi alle molecole--#
#Molecole dalla parte sinistra dell'uguale
print("Adesso dovrai inserire le molecole presenti dalla parte sinistra dell'uguale.")
MolecoleSinistra = []
while Stato != 'n': #Ciclo di creazione delle molecole all'interno di "MolecoleSinistra"
	Nome = input("Come si chiama questa molecola?")
	MolecoleSinistra += [Nome] #Il nome della molecola servirà a fornire l'output della risposta
	Contatore1 = 0
	while Contatore1 < len(Elementi) : #Ciclo di inserimento degli atomi in ogni molecola
		Elemento = input("Quanti atomi di " + Elementi[Contatore1] + " ha questa molecola?") #Riciclo della variabile "Elemento" per contare la quantita di elementi presente in ogni molecola
		MolecoleSinistra += [int(Elemento)]
		Contatore1 += 1
	Stato = input("C'è un'altra molecola da considerare? (S/n)")
print(MolecoleSinistra)

#Molecole dalla parte destra dell'uguale
print("Adesso dovrai inserire le molecole presenti dalla parte destra dell'uguale.")
Stato = 'S'
MolecoleDestra = []
while Stato != 'n': #Ciclo di creazione delle molecole all'interno di "MolecoleDestra"
	Nome = input("Come si chiama questa molecola?")
	MolecoleDestra += [Nome] #Il nome della molecola servirà a fornire l'output della risposta
	Contatore1 = 0
	while Contatore1 < len(Elementi): #Ciclo di inserimento degli atomi in ogni molecola
		Elemento = input("Quanti atomi di " + Elementi[Contatore1] + " ha questa molecola?") #Riciclo della variabile "Elemento" per contare la quantita di elementi presente in ogni molecola
		MolecoleDestra += [int(Elemento)]
		Contatore1 += 1
	Stato = input("C'è un'altra molecola da considerare? (S/n)")
print(MolecoleDestra)

##-------Creazione di funzioni e variabili che serviranno dopo--------##
#--Funzione che calcola il numero totale di elementi presente in una lista--#
def Prelevatore(NumeroElemento, lista):
	Contatore2 = 0
	Prelevatore = NumeroElemento
	ElementiTotali = 0
	while Prelevatore < len(lista):
		Elemento = lista[Prelevatore] #Prelevazione dell'elemento di una molecola dalla lista
		ElementiTotali += Elemento #Somma dell'elemento della molecola agli altri raccolti
		Contatore2 += 1
		Prelevatore = Contatore2 * (len(Elementi) + 1) + NumeroElemento #Passaggio ad un altra molecola
	return ElementiTotali

#--Variabili che rappresentano il numero di molecole contenuto dalle 2 parti dell'uguale--#
NumeroMolecoleSinistra = len(MolecoleSinistra) / (len(Elementi) + 1)
NumeroMolecoleDestra = len(MolecoleDestra) / (len(Elementi) + 1)

##-------------Effettuazione di tentativi di risoluzione--------------##
FattoreMassimo = int(input("Qual è il massimo coefficiente?"))

Stato = 'Tentativo fallito!'
while Stato == 'Tentativo fallito!': #Ciclo di generazione e validazione dei tentativi
	#--Generazione di fattori di molecole--#
	FattoriSinistra = []
	Contatore1 = 0
	while Contatore1 < NumeroMolecoleSinistra:
		FattoriSinistra += [random.randint(1, FattoreMassimo)]
		Contatore1 += 1
	FattoriDestra = []
	Contatore1 = 0
	while Contatore1 < NumeroMolecoleDestra:
		FattoriDestra += [random.randint(1, FattoreMassimo)]
		Contatore1 += 1
	#P.S. Mi rendo conto che questo metodo di generazione di fattori è assai rozzo, per questo sono ben accette idee migliori

	#--Creazione di molecole per testare il bilanciamento--#
	#Molecole dalla parte sinistra dell'uguale
	MolecoleVirtualiSinistra = [] #Molecole create per testare il bilanciamento elaborato precedentemente
	Contatore1 = 0
	while Contatore1 < NumeroMolecoleSinistra: #Ciclo di scrittura delle molecole
		MolecoleVirtualiSinistra += [MolecoleSinistra[Contatore1 * (len(Elementi) + 1)]] #Scrittura del nome della molecola nella molecola virtuale
		Contatore2 = 1
		Fattore = FattoriSinistra[Contatore1] #Prelievo del fattore di moltiplicazione della molecola
		while Contatore2 <= len(Elementi): #Ciclo di moltiplicazione di ogni molecola
			MolecoleVirtualiSinistra += [MolecoleSinistra[Contatore1 * (len(Elementi) + 1) + Contatore2] * Fattore] #Riscrittura della quantità dell'elemento moltiplicata per il fattore
			Contatore2 += 1 #Passaggio all'elemento successivo
		Contatore1 += 1 #Passaggio alla molecola successiva
	print(MolecoleVirtualiSinistra)

	#Molecole dalla parte destra dell'uguale
	MolecoleVirtualiDestra = [] #Molecole create per testare il bilanciamento elaborato precedentemente
	Contatore1 = 0
	while Contatore1 < NumeroMolecoleDestra: #Ciclo di scrittura delle molecole
		MolecoleVirtualiDestra += [MolecoleDestra[Contatore1 * (len(Elementi) + 1)]] #Scrittura del nome della molecola nella molecola virtuale
		Contatore2 = 1
		Fattore = FattoriDestra[Contatore1] #Prelievo del fattore di moltiplicazione della molecola
		while Contatore2 <= len(Elementi): #Ciclo di moltiplicazione di ogni molecola
			MolecoleVirtualiDestra += [MolecoleDestra[Contatore1 * (len(Elementi) + 1) + Contatore2] * Fattore] #Riscrittura della quantità dell'elemento moltiplicata per il fattore
			Contatore2 += 1 #Passaggio all'elemento successivo
		Contatore1 += 1 #Passaggio alla molecola successiva
	print(MolecoleVirtualiDestra)

	#--Validazione del lavoro prodotto--#
	Stato = 'Bilanciamento effettuato con successo!'
	Contatore1 = 1
	while Contatore1 <= len(Elementi): #Ciclo di controllo, elemento per elemento
		if Prelevatore(Contatore1, MolecoleVirtualiSinistra) != Prelevatore(Contatore1, MolecoleVirtualiDestra): #Confronto
			Stato = 'Tentativo fallito!'
		Contatore1 += 1 #Passaggio all'elemento successivo
	print(Stato)

#--Stampaggio su schermo della risposta--#
Contatore1 = 0
while Contatore1 < NumeroMolecoleSinistra:
	if(Contatore1 != 0):
		print('+ ', sep = '', end = "")
	print(FattoriSinistra[Contatore1], sep = '', end = "")
	print(MolecoleSinistra[Contatore1 * (len(Elementi) + 1)], sep = '', end = "")
	print(' ', sep = '', end = "")
	Contatore1 += 1
print('= ', end = "")
Contatore1 = 0
while Contatore1 < NumeroMolecoleDestra:
	print(FattoriDestra[Contatore1], sep = '', end = "")
	print(MolecoleDestra[Contatore1 * (len(Elementi) + 1)], sep = '', end = "")
	print(' ', sep = '', end = "")
	Contatore1 += 1

File: reactionBalancer_ITA.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void main(){
	char stato = 'n';
	int contatore1 = 1;
	int contatore2 = 0;
	int numeroElementi = 1;
	char nomiMolecoleSinistra[30][30];
	int molecoleSinistra[30][20];
	char nomiMolecoleDestra[30][30];
	int molecoleDestra[30][20];
	char elementi[30][2];
	int numeroMolecoleSinistra;
	int numeroMolecoleDestra;
	int coefficientiSinistra[30];
	int coefficientiDestra[30];
	int massimoCoefficiente;
	int molecoleVirtualiSinistra[30][20];
	int molecoleVirtualiDestra[30][20];
	int totSinistra;
	int totDestra;
	long int tentativi = 1;
	
	while(stato != 'S'){
		printf("Quale elemento è presente in questa reazione? ");
		scanf("%s", &elementi[numeroElementi]);
		printf("Questo era l'ultimo elemento? (S/n) ");
		scanf("%s", &stato);
		numeroElementi++;
	}
	printf("Ricapitolando, gli elementi che hai elencato sono: ");
	while(contatore1 < numeroElementi){
		printf("%s ", &elementi[contatore1]);
		contatore1++;
	}
	
	stato = 'S';
	contatore1 = 1;
	printf("\nAdesso dovrai inserire le molecole presenti dalla parte sinistra dell'uguale.\n");
	while(stato != 'n'){
		printf("Come si chiama questa molecola? ");
		scanf("%s", &nomiMolecoleSinistra[contatore1]);
		contatore2 = 1;
		while(contatore2 < numeroElementi){
			printf("Quanti atomi di %s ha questa molecola? ", &elementi[contatore2]);
			scanf("%d", &molecoleSinistra[contatore1][contatore2]);
			contatore2++;
		}
		contatore1++;
		printf("C'è un'altra molecola da considerare? (S/n) ");
		scanf("%s", &stato);
	}
	numeroMolecoleSinistra = contatore1;
	
	stato = 'S';
	contatore1 = 1;
	printf("Adesso dovrai inserire le molecole presenti dalla parte destra dell'uguale.\n");
	while(stato != 'n'){
		printf("Come si chiama questa molecola? ");
		scanf("%s", &nomiMolecoleDestra[contatore1]);
		contatore2 = 1;
		while(contatore2 < numeroElementi){
			printf("Quanti atomi di %s ha questa molecola? ", &elementi[contatore2]);
			scanf("%d", &molecoleDestra[contatore1][contatore2]);
			contatore2++;
		}
		contatore1++;
		printf("C'è un'altra molecola da considerare? (S/n) ");
		scanf("%s", &stato);
	}
	numeroMolecoleDestra = contatore1;
	
	printf("Qual è il massimo coefficiente? ");
	scanf("%d", &massimoCoefficiente);
	massimoCoefficiente--;
	
	stato = 'f';
	while(stato == 'f'){
		contatore1 = 1;
		while(contatore1 < numeroMolecoleSinistra){
			coefficientiSinistra[contatore1] = (rand() % massimoCoefficiente) + 1;
			contatore1++;
		}
		
		contatore1 = 1;
		while(contatore1 < numeroMolecoleDestra){
			coefficientiDestra[contatore1] = (rand() % massimoCoefficiente) + 1;
			contatore1++;
		}
		
		contatore1 = 1;
		while(contatore1 <= numeroMolecoleSinistra){
			contatore2 = 1;
			while(contatore2 < numeroElementi){
				molecoleVirtualiSinistra[contatore1][contatore2] = molecoleSinistra[contatore1][contatore2] * coefficientiSinistra[contatore1];
				contatore2++;
			}
			contatore1++;
		}
		
		contatore1 = 1;
		while(contatore1 <= numeroMolecoleDestra){
			contatore2 = 1;
			while(contatore2 < numeroElementi){
				molecoleVirtualiDestra[contatore1][contatore2] = molecoleDestra[contatore1][contatore2] * coefficientiDestra[contatore1];
				contatore2++;
			}
			contatore1++;
		}
		
		stato = 's';
		contatore1 = 1;
		while(contatore1 < numeroElementi){
			totSinistra = 0;
			contatore2 = 1;
			while(contatore2 <= numeroMolecoleDestra){
				totSinistra += molecoleVirtualiSinistra[contatore2][contatore1];
				contatore2++;
			}
			
			totDestra = 0;
			contatore2 = 1;
			while(contatore2 <= numeroMolecoleDestra){
				totDestra += molecoleVirtualiDestra[contatore2][contatore1];
				contatore2++;
			}
			
			if(totSinistra != totDestra){
				stato = 'f';
			}
			contatore1++;
		}
		
		contatore1 = 1;
		while(contatore1 < numeroMolecoleSinistra){
			printf("%d%s ", coefficientiSinistra[contatore1], &nomiMolecoleSinistra[contatore1]);
			contatore1++;
		}
		printf("= ");
		contatore1 = 1;
		while(contatore1 < numeroMolecoleDestra){
			printf("%d%s ", coefficientiDestra[contatore1], &nomiMolecoleDestra[contatore1]);
			contatore1++;
		}
		
		if(stato == 'f'){
			printf("Tentativo %d fallito\n", tentativi);
		}
		tentativi++;
	}
}
